shaderc-sys/build/build.rs

```diff
@@ -0,0 +1,194 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod cmd_finder;
+
+use std::env;
+use std::env::consts;
+use std::path::{Path, PathBuf};
+
+static COMBINED_LIB: &str = "shaderc_combined";
+static COMBINED_LIB_FILE: &str = "libshaderc_combined.a";
+static SPIRV_LIB: &str = "libSPIRV.a";
+
+fn build_shaderc(shaderc_dir: &PathBuf, use_ninja: bool) -> PathBuf {
+    let mut config = cmake::Config::new(shaderc_dir);
+    config
+        .profile("Release")
+        .define("CMAKE_POSITION_INDEPENDENT_CODE", "ON")
+        .define("SPIRV_SKIP_EXECUTABLES", "ON")
+        .define("SPIRV_WERROR", "OFF")
+        .define("SHADERC_SKIP_TESTS", "ON")
+        .define("CMAKE_INSTALL_LIBDIR", "lib");
+    if use_ninja {
+        config.generator("Ninja");
+    }
+    config.build()
+}
+
+fn build_shaderc_msvc(shaderc_dir: &PathBuf) -> PathBuf {
+    let mut config = cmake::Config::new(shaderc_dir);
+    config
+        .profile("Release")
+        .define("CMAKE_POSITION_INDEPENDENT_CODE", "ON")
+        .define("SPIRV_SKIP_EXECUTABLES", "ON")
+        .define("SPIRV_WERROR", "OFF")
+        .define("SHADERC_SKIP_TESTS", "ON")
+        // cmake-rs tries to be clever on Windows by injecting several
+        // C/C++ flags, which causes problems. So I have to manually
+        // define CMAKE_*_FLAGS_* here to suppress that.
+        .define("CMAKE_C_FLAGS", " /nologo /EHsc")
+        .define("CMAKE_CXX_FLAGS", " /nologo /EHsc")
+        .define("CMAKE_C_FLAGS_RELEASE", " /nologo /EHsc")
+        .define("CMAKE_CXX_FLAGS_RELEASE", " /nologo /EHsc")
+        .define("CMAKE_INSTALL_LIBDIR", "lib")
+        .generator("Ninja");
+    config.build()
+}
+
+fn main() {
+    // Deprecated --no-defaults path inhereted from before shaderc-rs & shaderc-sys split
+    if env::var("CARGO_FEATURE_CHECK_PASSTHROUGH").is_ok()
+        && env::var("CARGO_FEATURE_DONT_USE_DEPRECATED").is_err()
+    {
+        let out_dir = env::var("OUT_DIR").unwrap();
+        println!("cargo:warning=USE OF --no-defaults IS DEPRECATED BEHAVIOR.");
+        println!("cargo:warning=Requested to skip building native C++ shaderc.");
+        println!(
+            "cargo:warning=Searching {} for shaderc_combined static lib...",
+            out_dir
+        );
+        println!("cargo:rustc-link-search=native={}", out_dir);
+        println!("cargo:rustc-link-lib=static=shaderc_combined");
+        emit_stdc_link();
+        return;
+    }
+
+    // Explicit overrides first
+    let mut search_dir = if let Ok(lib_dir) = env::var("SHADERC_LIB_DIR") {
+        println!("cargo:warning=Using specified pre-built libshaderc");
+        Some(lib_dir)
+    } else if let Ok(lib_path) = env::var("SHADERC_LIB_PATH") {
+        println!("cargo:warning=Using specified pre-built libshaderc");
+        Some(lib_path)
+    } else if let Ok(lib_path) = env::var("SHADERC_STATIC") {
```

Comment:
Question: Why are we having three environment variables for the same functionality here?

Reply:
I'm going to change it to only use `SHADERC_LIB_DIR` since we might have multiple libraries and calling it a path would be false.

A lot of the choices attempted to defer to this document: https://kornel.ski/rust-sys-crate